geopandas/io/tests/test_sql.py

```diff
@@ -92,6 +82,79 @@ def engine_postgis():
     con.dispose()
 
 
+def prepare_psycopg_fixtures(fixture_prefix: Literal["connection", "engine"]):
+    """Work out which of psycopg2 and psycopg are available and create a parameter to
+    pass to pytest.mark.parametrize(), so tests run on none, one or both of the packages
+
+    This stops tests running if the relevant package isn't installed
+    (rather than being skipped, as skips are treated as failures during postgis CI)
+
+    Arguments:
+        fixture_prefix: "connection" or "engine", to match either name of either
+        connection_psycopg or engine_psycopg fixtures
+    """
+    connections_available = []
+    if find_spec("psycopg2"):
+        connections_available.append(f"{fixture_prefix}_psycopg2")
+    if find_spec("psycopg"):
+        connections_available.append(f"{fixture_prefix}_psycopg3")
+    return (
+        "parameterised_fixture",
+        connections_available,
+        True,
+    )
+
+
+CONNECTION_PARAMS = prepare_psycopg_fixtures("connection")
+ENGINE_PARAMS = prepare_psycopg_fixtures("engine")
+
+
+@pytest.fixture()
+def parameterised_fixture(request):
+    """A fixture that can accept other fixtures as a parameter.
+
+    Pass in the string name of the fixture needed."""
+    return request.getfixturevalue(request.param)
+
+
+@pytest.fixture
+def df_nybb(nybb_filename):
+    df = read_file(nybb_filename)
+    return df
+
+
+@pytest.fixture()
+def connection_psycopg2():
+    """
+    Use psycopg2 to initiate a connection to a postGIS database that must already exist.
+    See create_postgis for more information.
+    """
+    yield from connection_psycopg_any("psycopg2")
+
+
+@pytest.fixture()
+def connection_psycopg3():
+    """
+    Use psycopg3 to initiate a connection to a postGIS database that must already exist.
+    See create_postgis for more information.
+    """
+    yield from connection_psycopg_any(
+        "psycopg"
+    )  # package name is psycopg, not psycopg3
+
+
+@pytest.fixture()
+def engine_psycopg2():
+    """Uses psycopg2 to connect a sqlalchemy engine to an existing postGIS database."""
+    yield from engine_psycopg_any("psycopg2")
+
+
+@pytest.fixture()
+def engine_psycopg3():
```

Comment:
It'd be nice if we didn't need these engine/connection 2/3 variants but I don't think that's easy when we're hooking into the fixture setup time.

Reply:
Hmm, yes, they are pretty unclear and would be great to remove them. But this was the neatest I could come up with without creating a new engine/connection each time. I will have another look and see if there's a better way...